### ISSUE TITLE
Shortcut Ctrl+` not working when launching CopyQ with Czech keyboard on Windows

### ISSUE THREAD

**Issue by pbodnar** [hluk/CopyQ#1323]
**Describe the bug**
If I set a global keyboard shortcut like `` Ctrl+` `` for a command, it doesn't get mapped (i. e. it doesn't do anything) after Windows restarts. The same happens after user logout and login. But when I use a shortcut like `Ctrl+Alt+B`, there is no problem.

**Workaround (how to avoid the problem)**
1) Always start Windows / CopyQ with English,
2) don't forget to have English activated when assigning new shortcuts or when saving the CopyQ preferences. And "English" can be possibly exchanged for any other keyboard layout preferred by a user.

**To Reproduce**
1. Go to the CopyQ's preferences window, switch to the Shortcuts tab.
2. Assign shortcut `` Ctrl+` `` to the `Show/hide main window` command.
<img width="141" height="274" alt="Image" src="https://github.com/user-attachments/assets/63fd0ee9-f2d5-4505-aebc-1519999bce4a" />

3. Close the window.
4. Press the shortcut - the CopyQ's window appears = correct.
5. Switch to a different keyboard layout (e. g. from English to Czech).
6. Press the shortcut - the CopyQ's window disappears = correct.
7. Log off from Windows and log in again. UPDATE: It is enough just to quit CopyQ and start it again while non-English (Czech) keyboard layout is active.
8. Press the shortcut - the CopyQ's window doesn't appear = BUG.
**Expected behavior**
Any shortcut assigned to a command should work at any moment, including the moment after Windows restart.

**Screenshots**
n/a

**Version, OS and Environment**
 - Application Version [3.10.0]
 - OS [Windows 10] (with Czech locale)

**Additional context**
I guess that this might have something to do with a keyboard layouts "collision" - like that maybe mapping a pressed character works differently when CopyQ gets started automatically on Windows start. (There is also a weird Windows behavior that `Alt+Shift` won't switch keyboards until `Win+Space` is pressed or until an application is launched & focused.)

Note that there is also a historical bug #427 that coped with keyboard shortcuts and which might be related to this one.

**Comment by hluk** [comment]
Thanks for detailed description.

Can you also reproduce it by just restarting the app (Ctrl+Q and start again)? Maybe the shortcuts are not loaded properly at application start.

If it's a keyboard layout problem: Can you try assigning the shortcut when each layout is active? There is some [platform-specific key code translation](https://github.com/hluk/CopyQ/blob/e578111d96b6fe3e98668cb7e8725209f67c24dd/src/platform/win/winplatform.cpp#L337) that could cause the issue.

**Comment by pbodnar** [comment]
> Can you also reproduce it by just restarting the app (Ctrl+Q and start again)? Maybe the shortcuts are not loaded properly at application start.

That's it! :) Firstly, I wanted to write "sorry, I forgot to mention that starting the app manually is the 2nd workaround". But now that I have tested it again, the problem is unveiled:

1. Start the app when having Czech keyboard active -> shortcut fails. (equivalent of Windows start)
2. Start the app when having English keyboard active -> shortcut works.

**Conclusion: At application start the shortcuts loading is dependent on the current keyboard layout.**

_So a quick & dirty fix would be probably to run the same code as you run when re-saving the preferences dialog, just after you load the shortcuts at application start._

**Comment by hluk** [comment]
Hmm, so if you go to shortcut tab in preferences and press OK to apply (event thought there are no changes) the shortcuts work again?

Is there any warning or error in log (F12 shortcut from main window)?

**Comment by pbodnar** [comment]
Yes, exactly. It's 100% reproducible on my pc, but I can test it on my second pc on Monday for sure. 

In the log there is nothing special, only some "Connection lost" error messages from yesterday, no log items from today's testing, for example...

**Comment by pbodnar** [comment]
Update: Sorry for a confusion, but this is a really tricky bug: One needs to have English keyboard activated before clicking OK in the preferences. So when for example Czech is activated, the layout-dependent shortcut stops working. **No need for restarts to reproduce.**

**Comment by hluk** [comment]
Hmm, I think I would need to change [this code](https://github.com/hluk/CopyQ/blob/e578111d96b6fe3e98668cb7e8725209f67c24dd/src/platform/win/winplatform.cpp#L337), but I have no idea how.

Can you set multiple shortcuts for each layout? I know there could be potential conflicts but this is the only workaround I can think of -- unless you [set the shortcut in desktop environment](https://copyq.readthedocs.io/en/latest/faq.html#why-global-shortcuts-don-t-work).

**Comment by pbodnar** [comment]
> Hmm, I think I would need to change [this code](https://github.com/hluk/CopyQ/blob/e578111d96b6fe3e98668cb7e8725209f67c24dd/src/platform/win/winplatform.cpp#L337), but I have no idea how.

I'm not a Windows API expert myself, so I cannot help with this right now. I only slightly remember there was a possibility to listen for "virtual key" code (given by the key position on keyboard) instead for a translated character code...

> Can you set multiple shortcuts for each layout? I know there could be potential conflicts but this is the only workaround I can think of -- unless you [set the shortcut in desktop environment](https://copyq.readthedocs.io/en/latest/faq.html#why-global-shortcuts-don-t-work).

Yes, it looks like CopyQ lets you add a command multiple times and assign _differing keystrokes_ to the command instances, but that's not that practical as having just one keystroke to remember. 

Anyway, the **current workaround** doesn't luckily require too much hassle: 1) always start Windows / CopyQ with English, 2) don't forget to have English activated when assigning new shortcuts or when saving the CopyQ preferences. And "English" can be possibly exchanged for any other language preferred by a user.

**Comment by hluk** [comment]
> Yes, it looks like CopyQ lets you add a command multiple times and assign differing keystrokes to the command instances

You can assign multiple shortcuts to one command (in Command dialog or Shortcuts configuration tab in Preferences). Just click on the `+` button next to the shortcut.

![image](https://user-images.githubusercontent.com/38054/75801956-3ed9a400-5d7c-11ea-8ac0-3a95650f001a.png)

Not sure if it would work but you can try assigning the shortcut, then switch keyboard layout and assign the shortcut again (presumably GUI will show different shortcut).

**Comment by pbodnar** [comment]
Oh, that's cool, I haven't noticed this feature :) Still, CopyQ won't let me assign the same keystroke multiple times: the same keystroke is not accepted by the GUI even though I switch the layout as suggested. Tried from both, the Command dialog and Shortcuts configuration tab.

**Comment by Chealer** [comment]
Thank you very much for reporting @pbodnar 

~~I have no issues with <kbd>Ctrl</kbd>+<kbd>`</kbd> with version 13. Restarting CopyQ does not seem to change anything.~~

**Comment by Chealer** [comment]
I can reproduce this with CopyQ 13.0 with <kbd>Ctrl</kbd>+<kbd>\`</kbd> and <kbd>Shift</kbd>+<kbd>\`</kbd> if I start CopyQ with CAN/CSA Z243.200 (Canadian Multilingual Standard) as keyboard layout. This does not reproduce if I start in "French (Canada)" after having set the shortcuts in the legacy English QWERTY layout, even though these 2 layouts do not generate backquotes with the same keys.
It _looks_ like the shortcut does not get mapped, rather than simply getting mapped to different physical keys, but this is tricky to tell, since CAN/CSA Z243.200 needs a modifier to type backquotes. What is the situation with your layout? How is a backquote typed with the Czech layout?

@hluk wrote:
> Not sure if it would work but you can try assigning the shortcut, then switch keyboard layout and assign the shortcut again (presumably GUI will show different shortcut).

Thank you, this does work around for me.

**Comment by pbodnar** [comment]
@Chealer, thanks for your contribution to this topic.

> ... What is the situation with your layout? How is a backquote typed with the Czech layout?
>
> ...
>
> [@hluk](https://github.com/hluk) wrote:
> 
> > Not sure if it would work but you can try assigning the shortcut, then switch keyboard layout and assign the shortcut again (presumably GUI will show different shortcut).
> 
> Thank you, this does work around for me.

I've retested with CopyQ 9.1.0 and 13.0.0. And CopyQ already does let me assign the shortcut under 2 different keyboard layouts (pressing `` Ctrl+` `` (the top-left key under `Esc`) is presented as `` Ctrl+° `` by CopyQ _when Czech layout is activated_). Yet, the resulting behavior is still the same, i.e. the shortcut is not working when launching CopyQ with Czech layout activated.

(Anyway, I've recently switched to using `Win (Meta) + V`, because `` Ctrl+` `` was used by another program I started using.)

**Comment by Chealer** [comment]
Thank you @pbodnar 
Do you not find it odd that <kbd>Ctrl</kbd>+<kbd>`</kbd> becomes <kbd>Ctrl</kbd>+<kbd>°</kbd>? If the layout you are discussing is the one depicted [there](https://en.wikipedia.org/wiki/File:Windows_Czech_keyboard_layout.svg), should it not be a semi-colon (<kbd>;</kbd>)? I understand both characters are generated by the same key, but does "°" not belong to level 2?
Does this issue happen with any key other than that one?

<details>
  <summary>License</summary>
This comment and any other from myself on this ticket is offered under the terms of CC0 1.0.
</details>

**Comment by pbodnar** [comment]
@Chealer, you are right - it looks like <kbd>Ctrl</kbd>+<kbd>°</kbd> is the only case when "upper character" (<kbd>°</kbd>) is used instead of "lower character" (<kbd>;</kbd>) when pressing a key together with <kbd>Ctrl</kbd> in the CopyQ dialog. Maybe answering why this happens could lead us to finally resolving this issue?

**Comment by Chealer** [comment]
> [@Chealer](https://github.com/Chealer), you are right - it looks like Ctrl+° is the only case when "upper character" (°) is used instead of "lower character" (;) when pressing a key together with Ctrl in the CopyQ dialog.

Thank you

> Maybe answering why this happens could lead us to finally resolving this issue?

Yes, and knowing how shortcuts are stored too. The platform-specific code which Lukas referred to uses [the canonical `MapVirtualKeyW` function](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-mapvirtualkeyw), so I do not see a problem. It would also be good to check how other platforms behave.

Thank you for retitling, that is much clearer.

**Comment by Chealer** [comment]
@pbodnar: I'm sorry, but I no longer manage to really reproduce, with the same PC and CopyQ 14. It's not that it works, but I am struggling with step #2. Can you specify how you perform it, specifically entering the combination? Does it not require AltGr?

And is step #7 really necessary, despite what you wrote on 2020-03-01?

**Comment by pbodnar** [comment]
@Chealer, I just simply press <kbd>Ctrl</kbd>+<kbd>`</kbd> in the "New Shortcut" popup in step 2. I've rechecked with CopyQ 14 and it still all behaves the same.

Regarding step 7, you are right, I've updated the issue description.

**Comment by Chealer** [comment]
Thanks again @pbodnar 
I am sorry but I still fail to reproduce. I trust you that version 14 has not fixed, it's just that I must have forgotten how to perform step #2. Can you clarify how you “press <kbd>Ctrl</kbd>+<kbd>`</kbd>” with that layout? Which keys do you press?

**Comment by pbodnar** [comment]
@Chealer, OK, I've added an illustrative image of the keys to the description. :))

**Comment by Chealer** [comment]
Ooh, thank you @pbodnar . So if I understand correctly, you are not using the classic Czech QWERTZ or QWERTY layouts, but rather the programmers variant (KLID 00020405)?

**Comment by pbodnar** [comment]
> Ooh, thank you [@pbodnar](https://github.com/pbodnar) . So if I understand correctly, you are not using the classic Czech QWERTZ or QWERTY layouts, but rather the programmers variant (KLID 00020405)?

Nope, how did you infer it? I use plain English as the primary and plain Czech keyboard as the second keyboard. See below.

```plain
PS > Get-WinUserLanguageList

LanguageTag     : en-US
Autonym         : English (United States)
EnglishName     : English
LocalizedName   : English (United States)
ScriptName      : Latin
InputMethodTips : {0409:00000409}
Spellchecking   : True
Handwriting     : False

LanguageTag     : cs
Autonym         : čeština
EnglishName     : Czech
LocalizedName   : Czech
ScriptName      : Latin
InputMethodTips : {0405:00000405}
Spellchecking   : True
Handwriting     : False
```

**Comment by Chealer** [comment]
Thank you, but I cannot get a layout matching the one you pictured with 00000405. That is why I inferred you are using KLID 00020405; it is the only Czech keyboard layout I see which matches your screenshot. I am sorry this is so complicated, but are you still using Windows 10?
And does your keyboard's physical layout match the Czech software layout you use? If not, can you find a layout which matches your software layout on https://kbdlayout.info ?

**Comment by pbodnar** [comment]
@Chealer, you're welcome. But if you read carefully, there is _for example_ this step "5. Switch to a different keyboard layout (e. g. from English to Czech).". So logically, the screenshot I've added to step 2 was for the English layout.

Regarding what happens when I press the same key combo with the standard Czech layout (see [here](https://kbdlayout.info/kbdcz)), I've already described [here](https://github.com/hluk/CopyQ/issues/1323#issuecomment-3696149073). CopyQ thinks I want to assign the shortcut to a "dead key" (<kbd>°</kbd> instead to <kbd>;</kbd>), which apparently doesn't work correctly.

And yes, still on Windows 10. I can retest on Windows 11 next week, but I guess the results will be the same.

**Comment by Chealer** [comment]
Ooh, sorry Petr, I was indeed confused. Thanks for clarifying

@pbodnar wrote:
> And yes, still on Windows 10.

OK, thanks

> I can retest on Windows 11 next week, but I guess the results will be the same.

Thanks but don't bother, not needed

I think I am confused by the title and the body. Is the _Describe the bug_ section accurate? Can you reliably reproduce while just restarting Windows, without changing layout?

You are most **definitely** reporting a bug in CopyQ; but I feel the description is incorrect. Despite what I wrote above, I technically cannot reproduce with your current steps; if I switch to [CAN/CSA Z243.200-92 ("Canadian Multilingual Standard")](https://en.wikipedia.org/wiki/CSA_keyboard) at step 5, there is no effect at step 6.
As for Czech, can you specify step 6? Which exact keypresses do you do there to reproduce in Czech?

And can you find other keyboard layouts than Czech with which this reproduces?

**Comment by hluk** [comment]
I'm not sure if this can be fixed in any reasonable way. The app uses Qt to capture shortcuts (QKeySequenceEdit) which relies on the current keyboard layout - i.e. storing logical key codes instead physical ones.

The relevant shortcut conversion code for Windows now also works with current keyboard layout (it just falls back to US layout - which should be unnecessary): https://github.com/hluk/CopyQ/blob/4fdd9d986a9f63769917b952f2c44c8aaa55c5d8/qxt/qxtglobalshortcut_win.cpp#L206-L213

So I think you would need to **configure a shortcut for each keyboard layout you use**.

**Comment by pbodnar** [comment]
I have just updated the issue description, so that hopefully all the discussed major findings are incorporated there.

@hluk, so in another words, are you saying the behavior is by-design and/or limited by how the Qt library works? And what about the keyboard layout _when CopyQ starts_? It apparently has an impact on how the pressed keys are recognized, no matter which keyboard layout is _currently_ active.

**Comment by hluk** [comment]
> so in another words, are you saying the behavior is by-design and/or limited by how the Qt library works?

I never put much thought into how global shortcuts are supposed to work with multiple layouts, but I guess they supposed to work only with specific keyboard layout - most of the shortcuts also show in specific layout in UI. Or am I mistaken?

> And what about the keyboard layout _when CopyQ starts_? It apparently has an impact on how the pressed keys are recognized, no matter which keyboard layout is _currently_ active.

Oh, I think this could be a bug. The global shortcuts a registered in current layout when app starts or configuration changes.

**Comment by Chealer** [comment]
Thank you Lukas

> > so in another words, are you saying the behavior is by-design and/or limited by how the Qt library works?
> 
> I never put much thought into how global shortcuts are supposed to work with multiple layouts, but I guess they supposed to work only with specific keyboard layout - most of the shortcuts also show in specific layout in UI. Or am I mistaken?

Qt's shortcut handling is a mess, and even more on Windows. [Qt itself doesn't really know how it behaves](https://qt-project.atlassian.net/browse/QTBUG-112736).

I recommend putting a warning about custom shortcuts being unreliable, at least on Windows, perhaps pointing to [QTBUG-116873](https://qt-project.atlassian.net/browse/QTBUG-116873). We could suggest to use only modifiers and non-diacriticized letters until Qt is cleaned up.

**Comment by Chealer** [comment]
Thank you for the reply and for the update @pbodnar, but that does not answer my questions, in particular about how you perform step #6.

> There is also a historical bug [#427](https://github.com/hluk/CopyQ/issues/427) that coped with keyboard shortcuts and which might be related to this one.

Not really. Issue #427 is unfortunately not "historical", but it does not require switching layout.

**Comment by hluk** [comment]
Can you try [this development build](https://github.com/hluk/CopyQ/actions/runs/24066621436?pr=3536#artifacts)? It should re-register global shortcuts whenever the keyboard layout changes.

**Comment by pbodnar** [comment]
@Chealer, what about now when I've updated the description again?

@hluk, thanks for the build. I have tested it and I'm not quite sure whether it makes the situation better though, because:

1. CopyQ seems to realize the layout was switched only when it has focus. So the CopyQ's response to defined shortcuts can differ "in time".
2. The problem with invalid recognition of `` Ctrl+` `` as `` Ctrl+° `` for Czech layout still persists. As @Chealer usefully pointed out, this might be a problem in Qt itself. I've noticed _for example_ this issue in their tracker: [QTBUG-90611 - Shortcut Alt+` is not working with non US layout](https://qt-project.atlassian.net/browse/QTBUG-90611).

But I've found another workaround: assigning `` Ctrl+; `` as another (and actually apparently correct) shortcut for the case Czech layout is active. This can be done while having English layout active in the shortcut creation popup, or by entering it into `config\copyq\copyq-commands.ini` directly (surrounded by quotes probably because of the ini file syntax):

```ini
1\Command=copyq: toggle()
1\IsGlobalShortcut=true
1\GlobalShortcut=meta+v, "ctrl+;", ctrl+`
...
```

**Comment by Chealer** [comment]
Thanks Petr, but my question was about step #6. Step #4 was already clear enough to my taste.
I doubt you would experience QTBUG-90611 on the Windows build of CopyQ 13.0, which uses a Qt much more recent than the first one which reportedly has that fixed. It seems more likely you would experience QTBUG-112736 or [QTBUG-111992](https://qt-project.atlassian.net/browse/QTBUG-111992).

@hluk: I am confused by your change. I both read the message and glanced over the code, and my understanding is this targets the moment when the layout changes. But according to the reproduction steps, the layout change does not directly cause the bug, it's (re)launching CopyQ which triggers.
